File: rs2-core/src/services/pipeline_service.rs
//! `pipeline` — a pipeline store (PRD §10.3), v1's store-transform pattern.
//!
//! Authoring lives under the reserved subtree `/<mount>/.pipelines/…` — a
//! normal store-contract surface delegated to the owned `FileService` via
//! [`SpecStore`], with envelopes validated (and DSL canonicalized to the
//! typed spec) at write time. `GET <spec>?$plan` returns the segment plan.
//!
//! Every other path, on **any HTTP verb**, executes: the longest stored
//! prefix wins, and a spec named `.root` governs the mount root — so a
//! pipeline can transparently wrap another service (custom security
//! context, unchanged API), the reason pipelines pass verbs through.

use async_trait::async_trait;
use serde_json::Value;

use super::spec_store::SpecStore;
use super::{Service, ServiceContext};
use crate::error::RsError;
use crate::message::{Message, Source};
use crate::pipeline::{plan, Executor, PipelineLimits, PipelineSpec};
use crate::retry::RetryPolicy;

/// Default storage prefix (under the tenant file store) for pipeline mounts.
pub const PIPELINE_PREFIX: &str = ".rs2-pipelines";

/// The reserved authoring subtree segment.
pub const PIPELINE_SUBTREE: &str = ".pipelines";

pub struct PipelineService {
    store: SpecStore,
}

impl PipelineService {
    pub fn from_config(config: &Value, store: SpecStore) -> Result<Self, RsError> {
        if config.get("pipeline").is_some() {
            return Err(RsError::bad_request(
                "config-defined pipelines are no longer supported: PUT the spec envelope to \
                 /<mount>/.pipelines/<name> (or .pipelines/.root to govern the mount root)",
            ));
        }
        Ok(PipelineService { store })
    }

    /// Write-time validator: envelope `{pipeline, retry?, description?,
    /// x-…}`; the pipeline (typed or string DSL) is converted and validated,
    /// and the **typed form** is what gets stored (the PRD's stored format).
    pub fn validator() -> super::spec_store::SpecValidator {
        std::sync::Arc::new(|doc: &Value| {
            let obj = doc.as_object().ok_or_else(|| {
                RsError::bad_request("a stored pipeline is a JSON object envelope")
            })?;
            let spec_value = obj
                .get("pipeline")
                .ok_or_else(|| RsError::bad_request("pipeline envelope requires a 'pipeline'"))?;
            let spec = PipelineSpec::from_value(spec_value)?;
            if let Some(retry) = obj.get("retry") {
                serde_json::from_value::<RetryPolicy>(retry.clone())
                    .map_err(|e| RsError::bad_request(format!("invalid 'retry' policy: {e}")))?;
            }
            if let Some(access) = obj.get("access") {
                validate_access_shape(access)?;
            }
            let mut canonical = obj.clone();
            canonical.insert(
                "pipeline".to_string(),
                serde_json::to_value(&spec).map_err(|e| RsError::internal(e.to_string()))?,
            );
            Ok(Value::Object(canonical))
        })
    }

    fn spec_from_doc(doc: &Value) -> Result<PipelineSpec, RsError> {
        serde_json::from_value(doc.get("pipeline").cloned().unwrap_or(Value::Null))
            .map_err(|e| RsError::internal(format!("stored pipeline is corrupt: {e}")))
    }
}

/// The effective per-spec access: the spec's `access` overlaid on the mount's
/// `access` floor. When both are role objects the spec wins **per key** (so a
/// spec can loosen `invoke` while inheriting the mount's `read`); otherwise the
/// most-specific present value (the spec) wins wholesale. `None` ⇒ open.
fn effective_access(mount: Option<&Value>, spec: Option<&Value>) -> Option<Value> {
    match (mount, spec) {
        (None, None) => None,
        (Some(m), None) => Some(m.clone()),
        (None, Some(s)) => Some(s.clone()),
        (Some(Value::Object(m)), Some(Value::Object(s))) => {
            let mut merged = m.clone();
            for (k, v) in s {
                merged.insert(k.clone(), v.clone());
            }
            Some(Value::Object(merged))
        }
        (Some(_), Some(s)) => Some(s.clone()),
    }
}

/// Validate the shape of a spec envelope's `access` field: a string shorthand
/// (`"open"` / `"authenticated"`) or an object of string role specs keyed only
/// by the action vocabulary (`read`/`write`/`delete`/`invoke`). `manage` and
/// unknown keys are rejected (a spec is not a management scope, and the guard
/// catches typos like `invokeRoles`).
fn validate_access_shape(access: &Value) -> Result<(), RsError> {
    match access {
        Value::String(s) if s == "open" || s == "authenticated" => Ok(()),
        Value::String(s) => Err(RsError::bad_request(format!(
            "unknown access policy '{s}' (expected \"open\" or \"authenticated\")"
        ))),
        Value::Object(map) => {
            for (key, val) in map {
                if !matches!(key.as_str(), "read" | "write" | "delete" | "invoke") {
                    return Err(RsError::bad_request(format!(
                        "unknown access key '{key}' (allowed: read, write, delete, invoke)"
                    )));
                }
                if !val.is_string() {
                    return Err(RsError::bad_request(format!(
                        "access '{key}' must be a role-spec string"
                    )));
                }
            }
            Ok(())
        }
        _ => Err(RsError::bad_request(
            "'access' must be a string or role object",
        )),
    }
}

#[async_trait]
impl Service for PipelineService {
    async fn handle(&self, msg: Message, ctx: &ServiceContext) -> Result<Message, RsError> {
        if self.store.is_authoring(&msg) {
            // `GET <spec>?$plan` — segment-plan introspection (PRD §8.3).
            if msg.method == http::Method::GET
                && !msg.url.is_directory()
                && msg.url.query_param("$plan").is_some()
            {
                let rel = msg
                    .url
                    .service_path
                    .strip_prefix(&format!("/{PIPELINE_SUBTREE}"))
                    .unwrap_or("/")
                    .to_string();
                let doc = self.store.read(&rel).await?;
                let spec = Self::spec_from_doc(&doc)?;
                return Ok(msg.ok_json(&serde_json::json!({
                    "pipeline": spec,
                    "plan": plan(&spec),
                })));
            }
            return self.store.handle_authoring(msg).await;
        }

        // ---- execution: any verb, longest stored prefix, .root fallback ----
        let segments: Vec<String> = msg
            .url
            .service_segments()
            .iter()
            .map(|s| s.to_string())
            .collect();
        let Some((doc, matched_len)) = self.store.resolve(&segments).await? else {
            return Err(RsError::not_found(format!(
                "no stored pipeline matches '{}' (author one at {}{}/…)",
                msg.url.service_path, msg.url.base_path, PIPELINE_SUBTREE,
            )));
        };
        let spec = Self::spec_from_doc(&doc)?;

        // The peeled sub-path (segments beyond the matched spec prefix) is the
        // URL plane for `${url.path[…]}` in call URLs — so a `.root` spec can
        // transparently forward the addressed key (e.g. `/users/<email>` →
        // `/data/users/${url.path[0]}`).
        let peeled: Vec<String> = segments[matched_len..].to_vec();
        let base_segs: Vec<String> = msg
            .url
            .base_segments()
            .iter()
            .map(|s| s.to_string())
            .collect();
        let url_name = peeled.last().cloned();
        let url_query = msg.url.query.clone();

        // Per-spec authorization. The host defers a pipeline mount's execution
        // surface to here: the matched spec's `access` overrides the mount's
        // `access` floor per key (`.root` is the mount-wide floor), evaluated
        // with the same verb→action map — POST→invoke = "run this pipeline".
        //
        // Fail closed: with no `access` on either the mount or the matched spec,
        // execution is denied (401 anonymous / 403 authenticated) — never run
        // open by default. A public pipeline must opt in with `"access": "open"`
        // on the mount or the spec.
        match effective_access(ctx.config.get("access"), doc.get("access")) {
            Some(access) => crate::wrapper::check_role_spec(
                &access,
                crate::wrapper::action_for(&msg.method),
                &msg,
            )?,
            // A runtime-originated system call (a scheduler tick) is trusted —
            // it exists only by operator config and can't be forged from the
            // wire — matching `check_role_spec`'s trust of `Source::System`.
            None if msg.source == Source::System => {}
            None if msg.principal.is_some() => {
                return Err(RsError::forbidden(
                    "this pipeline has no access policy configured",
                ))
            }
            None => {
                return Err(RsError::unauthorized(
                    "this pipeline has no access policy configured",
                ))
            }
        }

        let requester = ctx
            .requester
            .clone()
            .ok_or_else(|| RsError::internal("pipeline service has no requester capability"))?;
        let to_step = msg
            .url
            .query_param("$to-step")
            .and_then(|v| v.parse::<usize>().ok());

        let limits = PipelineLimits {
            wall_clock: ctx.pipeline_wall_clock,
            materialize_cap: ctx.limits.materialized_body_bytes,
            ..PipelineLimits::default()
        };
        // Retry resolution: envelope → mount config → tenant default.
        let envelope_retry = doc.get("retry").and_then(RetryPolicy::from_config);
        let mount_retry = RetryPolicy::from_config(ctx.config.get("retry").unwrap_or(&Value::Null));
        let retry = RetryPolicy::resolve(&[
            envelope_retry.as_ref(),
            mount_retry.as_ref(),
            ctx.tenant_retry.as_ref(),
        ]);
        let mut executor = Executor::new(requester, limits.clone(), retry)
            // The role an `elevate` step adds, from operator-controlled mount
            // config (`"elevate": "<role>"`). Authority lives here, not in the
            // authored spec.
            .with_elevate_role(
                ctx.config
                    .get("elevate")
                    .and_then(|v| v.as_str())
                    .map(str::to_string),
            )
            .with_url(peeled, base_segs, url_name, url_query);
        // Bind the mount's granted secrets as `$<name>` variables (host-side),
        // so a transform can `$hmacVerify('sha256', $<name>, $_rawBody, $sig)`.
        if let Some(secrets) = &ctx.secrets {
            executor = executor.with_vars(secrets.clone());
        }

        let result = match tokio::time::timeout(
            limits.wall_clock,
            executor.run(&spec, msg, to_step),
        )
        .await
        {
            Ok(result) => result,
            Err(_) => Err(RsError::limit_exceeded(
                "pipeline_wall_clock_ms",
                limits.wall_clock.as_millis() as u64,
                limits.wall_clock.as_millis() as u64,
            )),
        };

        // Pipeline failures include the failing step and per-step statuses
        // (PRD §12) so agents can recover instead of guessing.
        match result {
            Ok(mut resp) if !resp.is_ok() => {
                let steps = executor.report();
                let failed = steps
                    .iter()
                    .rev()
                    .find(|s| s["status"].as_u64().unwrap_or(0) >= 400)
                    .cloned();
                if let Some(body) = &mut resp.body {
                    if body.media_type.is_json() {
                        if let Ok(mut problem) =
                            body.as_json(ctx.limits.materialized_body_bytes).await
                        {
                            if let Some(obj) = problem.as_object_mut() {
                                obj.insert(
                                    "pipeline".to_string(),
                                    serde_json::json!({
                                        "failedStep": failed.as_ref().and_then(|f| f.get("step")),
                                        "steps": steps,
                                    }),
                                );
                                let media_type = body.media_type.clone();
                                resp.body = Some(crate::message::Body::from_string(
                                    problem.to_string(),
                                    media_type,
                                ));
                            }
                        }
                    }
                }
                Ok(resp)
            }
            other => other,
        }
    }
}

File: rs2-worker/src/services/pipeline-service.ts
// `pipeline` — a pipeline store (PRD §10.3), v1's store-transform pattern.
// Authoring lives under `/<mount>/.pipelines/…` (a `SpecStore` over the
// owned file service, envelopes validated and DSL canonicalized at write
// time; `GET <spec>?$plan` returns the segment plan). Every other path, on
// any verb, executes: the longest stored prefix wins, `.root` governs the
// mount root. Port of `rs2-core/src/services/pipeline_service.rs`.

import { Executor, defaultPipelineLimits } from "../pipeline/executor";
import { convert as convertDsl } from "../pipeline/dsl";
import { plan, planToJson } from "../pipeline/segments";
import { specFromJson, specFromValue, specToJson } from "../pipeline/spec";
import type { PipelineSpec } from "../pipeline/spec";
import { Body } from "../runtime/body";
import { RsError } from "../runtime/error";
import type { Json, JsonObject } from "../runtime/error";
import { MediaType } from "../runtime/media-type";
import type { Message } from "../runtime/message";
import { ExternalDispatch, isExternalUrl, urlHost } from "../runtime/outbound";
import { resolveRetry, retryFromConfig } from "../runtime/retry";
import type { RetryPolicy } from "../runtime/retry";
import { actionFor, checkRoleSpec } from "../runtime/wrapper";
import { PIPELINE_SUBTREE } from "./context";
import type { Service, ServiceContext } from "./context";
import type { SpecStore, SpecValidator } from "./spec-store";

export class PipelineService implements Service {
  private constructor(private readonly store: SpecStore) {}

  static fromConfig(config: JsonObject, store: SpecStore): PipelineService {
    if (config.pipeline !== undefined) {
      throw RsError.badRequest(
        "config-defined pipelines are no longer supported: PUT the spec envelope to /<mount>/.pipelines/<name> (or .pipelines/.root to govern the mount root)",
      );
    }
    return new PipelineService(store);
  }

  /// Write-time validator: envelope `{pipeline, retry?, description?, x-…}`;
  /// the pipeline (typed or string DSL) is converted and validated, and the
  /// typed form is what gets stored.
  static validator(): SpecValidator {
    return (doc: Json): Json => {
      if (!doc || typeof doc !== "object" || Array.isArray(doc)) {
        throw RsError.badRequest("a stored pipeline is a JSON object envelope");
      }
      const specValue = doc.pipeline;
      if (specValue === undefined) throw RsError.badRequest("pipeline envelope requires a 'pipeline'");
      const spec = specFromValue(specValue, convertDsl);
      if (doc.retry !== undefined) {
        if (doc.retry !== null && !retryFromConfig(doc.retry)) {
          throw RsError.badRequest("invalid 'retry' policy: expected a retry policy object");
        }
      }
      if (doc.access !== undefined) validateAccessShape(doc.access);
      const canonical: JsonObject = { ...doc, pipeline: specToJson(spec) };
      return canonical;
    };
  }

  private static specFromDoc(doc: Json): PipelineSpec {
    const value = doc && typeof doc === "object" && !Array.isArray(doc) ? (doc.pipeline ?? null) : null;
    try {
      return specFromJson(value);
    } catch (e) {
      throw RsError.internal(`stored pipeline is corrupt: ${(e as RsError).detail ?? String(e)}`);
    }
  }

  async handle(msg: Message, ctx: ServiceContext): Promise<Message> {
    if (this.store.isAuthoring(msg)) {
      // `GET <spec>?$plan` — segment-plan introspection (PRD §8.3).
      if (msg.method === "GET" && !msg.url.isDirectory() && msg.url.queryParam("$plan") !== undefined) {
        const prefix = `/${PIPELINE_SUBTREE}`;
        const rel = msg.url.servicePath.startsWith(prefix) ? msg.url.servicePath.slice(prefix.length) : "/";
        const doc = await this.store.read(rel === "" ? "/" : rel);
        const spec = PipelineService.specFromDoc(doc);
        const segPlan = plan(spec);
        // Static external-coverage check: literal absolute call URLs whose
        // host no `httpOut` grant covers would be denied at execution.
        const external = ExternalDispatch.fromMount(ctx.config, ctx.http, ctx.outboundInjectors, ctx.limits.materializedBodyBytes);
        appendExternalWarnings(spec, "", external, segPlan.warnings);
        return msg.okJson({ pipeline: specToJson(spec), plan: planToJson(segPlan) });
      }
      return this.store.handleAuthoring(msg);
    }

    // ---- execution: any verb, longest stored prefix, .root fallback ----
    const segments = msg.url.serviceSegments();
    const resolved = await this.store.resolve(segments);
    if (!resolved) {
      throw RsError.notFound(
        `no stored pipeline matches '${msg.url.servicePath}' (author one at ${msg.url.basePath}${PIPELINE_SUBTREE}/…)`,
      );
    }
    const [doc, matchedLen] = resolved;
    const spec = PipelineService.specFromDoc(doc);

    // The peeled sub-path is the URL plane for `${url.path[…]}`.
    const peeled = segments.slice(matchedLen);
    const baseSegs = msg.url.baseSegments();
    const urlName = peeled.length ? peeled[peeled.length - 1] : undefined;
    const urlQuery = msg.url.query;

    // Per-spec authorization: the matched spec's `access` overrides the
    // mount's floor per key. Fail closed when neither declares one.
    const docObj = doc && typeof doc === "object" && !Array.isArray(doc) ? doc : {};
    const access = effectiveAccess(ctx.config.access, docObj.access);
    if (access !== undefined) {
      checkRoleSpec(access, actionFor(msg.method), msg);
    } else if (msg.source === "system") {
      /* a runtime-originated system call (a scheduler tick) is trusted */
    } else if (msg.principal) {
      throw RsError.forbidden("this pipeline has no access policy configured");
    } else {
      throw RsError.unauthorized("this pipeline has no access policy configured");
    }

    const envelopeRetry = docObj.retry !== undefined ? retryFromConfig(docObj.retry) : undefined;
    const rest = rebuildRest(peeled, msg.url.isDirectory());
    return runInline(spec, msg, ctx, { peeled, baseSegs, urlName, urlQuery, rest, envelopeRetry });
  }
}

/// The effective per-spec access: the spec's `access` overlaid on the
/// mount's floor. Both role objects ⇒ the spec wins per key; otherwise the
/// most-specific present value wins wholesale. `undefined` ⇒ open.
function effectiveAccess(mount: Json | undefined, spec: Json | undefined): Json | undefined {
  if (mount === undefined && spec === undefined) return undefined;
  if (spec === undefined) return mount;
  if (mount === undefined) return spec;
  if (mount && typeof mount === "object" && !Array.isArray(mount) && spec && typeof spec === "object" && !Array.isArray(spec)) {
    return { ...mount, ...spec };
  }
  return spec;
}

/// Validate the shape of a spec envelope's `access` field.
function validateAccessShape(access: Json): void {
  if (typeof access === "string") {
    if (access === "open" || access === "authenticated") return;
    throw RsError.badRequest(`unknown access policy '${access}' (expected "open" or "authenticated")`);
  }
  if (access && typeof access === "object" && !Array.isArray(access)) {
    for (const [key, val] of Object.entries(access)) {
      if (!["read", "write", "delete", "invoke"].includes(key)) {
        throw RsError.badRequest(`unknown access key '${key}' (allowed: read, write, delete, invoke)`);
      }
      if (typeof val !== "string") throw RsError.badRequest(`access '${key}' must be a role-spec string`);
    }
    return;
  }
  throw RsError.badRequest("'access' must be a string or role object");
}

/// URL-plane inputs for `runInline` — the incoming request's path, peeled
/// to the plane a pipeline addresses, plus the verbatim `rest`.
export interface ExecInputs {
  peeled: string[];
  baseSegs: string[];
  urlName: string | undefined;
  urlQuery: string;
  rest: string;
  envelopeRetry: RetryPolicy | undefined;
}

/// Build the executor with the request's URL plane, run `spec` under the
/// wall-clock budget, and shape a failure into a structured problem with
/// per-step statuses (PRD §12). Shared by `pipeline` and `wrapper`.
export async function runInline(spec: PipelineSpec, msg: Message, ctx: ServiceContext, inputs: ExecInputs): Promise<Message> {
  const requester = ctx.requester;
  if (!requester) throw RsError.internal("pipeline service has no requester capability");
  const toStepRaw = msg.url.queryParam("$to-step");
  const toStep = toStepRaw !== undefined && /^\d+$/.test(toStepRaw) ? Number(toStepRaw) : undefined;

  const limits = { ...defaultPipelineLimits(), wallClockMs: ctx.pipelineWallClockMs, materializeCap: ctx.limits.materializedBodyBytes };
  // Retry resolution: envelope → mount config → tenant default.
  const mountRetry = retryFromConfig(ctx.config.retry ?? null);
  const retry = resolveRetry([inputs.envelopeRetry, mountRetry, ctx.tenantRetry]);
  const elevate = ctx.config.elevate;
  const executor = new Executor(requester, limits, retry)
    .withElevateRole(typeof elevate === "string" ? elevate : undefined)
    .withExternal(ExternalDispatch.fromMount(ctx.config, ctx.http, ctx.outboundInjectors, ctx.limits.materializedBodyBytes))
    .withUrl(inputs.peeled, inputs.baseSegs, inputs.urlName, inputs.urlQuery, inputs.rest);
  if (ctx.secrets) executor.withVars({ ...ctx.secrets });

  let timer: ReturnType<typeof setTimeout> | undefined;
  const timeout = new Promise<never>((_, reject) => {
    timer = setTimeout(
      () => reject(RsError.limitExceeded("pipeline_wall_clock_ms", limits.wallClockMs, limits.wallClockMs)),
      limits.wallClockMs,
    );
  });
  let resp: Message;
  try {
    resp = await Promise.race([executor.run(spec, msg, toStep), timeout]);
  } finally {
    if (timer !== undefined) clearTimeout(timer);
  }

  // Pipeline failures include the failing step and per-step statuses.
  if (!resp.isOk() && resp.body && resp.body.mediaType.isJson()) {
    const steps = executor.report();
    const failed = [...steps].reverse().find((s) => typeof s.status === "number" && s.status >= 400);
    let problem: Json | undefined;
    try {
      problem = await resp.body.asJson(ctx.limits.materializedBodyBytes);
    } catch {
      problem = undefined;
    }
    if (problem && typeof problem === "object" && !Array.isArray(problem)) {
      problem.pipeline = { failedStep: failed?.step ?? null, steps };
      const mediaType = resp.body.mediaType;
      resp.body = Body.fromString(JSON.stringify(problem), new MediaType(mediaType.essence(), mediaType.schema()));
    }
  }
  return resp;
}

/// `?$plan` static coverage: warn on literal external call URLs whose host
/// no `httpOut` grant on the mount covers.
function appendExternalWarnings(spec: PipelineSpec, at: string, external: ExternalDispatch | undefined, warnings: string[]): void {
  spec.steps.forEach((step, i) => {
    if (step.call && isExternalUrl(step.call.url)) {
      const host = urlHost(step.call.url);
      if (host !== undefined && !host.includes("${") && !(external?.covers(host) ?? false)) {
        warnings.push(
          `${at}/steps[${i}]: external host '${host}' is not covered by any httpOut grant on this mount — the call would be denied`,
        );
      }
    }
    if (step.pipeline) appendExternalWarnings(step.pipeline, `${at}/steps[${i}]/pipeline`, external, warnings);
  });
}

/// Rebuild the `${url.rest}` value from peeled segments: a leading slash,
/// the segments `/`-joined, and a trailing slash for a directory request.
export function rebuildRest(peeled: string[], isDirectory: boolean): string {
  if (peeled.length === 0) return "/";
  return `/${peeled.join("/")}${isDirectory ? "/" : ""}`;
}
